### PR TITLE
[UENV] CP2K: compilation and performance tests

### PR DESCRIPTION
**Pull request by RMeli**
CP2K compilation with the `develop` view, and performance tests for the executable provided by the UENV.

* [x] Parametrise compilation test on the version in the UENV
* [ ] ~Explore integration of full unit tests~
* [x] Compilation with `modules` view?

### CODE REVIEW

File: checks/apps/cp2k/cp2k_check.py [hunk 1]
```diff
@@ -3,149 +3,227 @@
...
+
+    def __init__(self):
+        super().__init__()
+
+    @run_before("run")
+    def prepare_run(self):
+        super().prepare_run()
+        self.job.options += [
+            "--time 00:05:00",
+        ]
```
Comment (RMeli) [inline checks/apps/cp2k/cp2k_check.py:201]: This doesn't work...
```
[...]
#SBATCH --time 0:10:0
[...]
#SBATCH --time 00:15:00
[...]
```
The default time is still present.

File: checks/apps/cp2k/cp2k_check.py [hunk 2]
```diff
@@ -3,149 +3,208 @@
...
+            self.build_system.config_opts += [
+                "-DCP2K_USE_ACCEL=CUDA",
+                "-DCP2K_WITH_GPU=H100",
+            ]
+
+    @sanity_function
+    def validate_test(self):
+        # INFO: Executables are in exe/local_cuda because -DCP2K_ENABLE_REGTEST=ON
+        # INFO: With -DCP2K_ENABLE_REGTEST=OFF, executables are in build/bin/ as expected
+        executable = os.path.join(self.stagedir, "exe", "local_cuda", "cp2k.psmp")
```
Comment (RMeli) [inline checks/apps/cp2k/cp2k_check.py:102]: Folder `local_cuda` needs to be conditional on `uarch`.

File: checks/apps/cp2k/cp2k_check.py [hunk 3]
```diff
@@ -3,149 +3,208 @@
...
-        self.reference = self.refs_by_scale[self.scale]
-
-    @run_before('run')
-    def set_task_distribution(self):
-        self.job.options = ['--distribution=block:block']
-
-    @run_before('run')
-    def set_cpu_binding(self):
-        self.job.launcher.options = ['--cpu-bind=cores']
+class Cp2kCheckPBE(Cp2kCheck):
```
Comment (RMeli) [inline checks/apps/cp2k/cp2k_check.py:184]: Run PBE test with built executable too? `+cp2k-dev` environment feature.

**Comment by RMeli** [comment]
- [ ] Add `README.md` with file provenance
- [ ] Implement dependencies in a cleaner way
- [ ] Use shorter tests

File: checks/apps/cp2k/cp2k_check.py [hunk 4]
```diff
@@ -3,149 +3,208 @@
...
+import os
+import shutil
+
 import reframe as rfm
 import reframe.utility.sanity as sn
+import reframe.utility.udeps as udeps
+import uenv
+
+cp2k_references = {
+    "pbe": {"gh200": {"time_run": (65, None, 5, "s")}},
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:15]: Do you expect an upper limit of 65*5 or reference +5% ?
```suggestion
    "pbe": {"gh200": {"time_run": (65, None, 5/100, "s")}},
```

File: checks/apps/cp2k/src/BASIS_H2O [hunk 5]
```diff
@@ -0,0 +1,244 @@
+#
```
Comment (jgphpc) [inline checks/apps/cp2k/src/BASIS_H2O:1]: add a readme.md to show the origin of the input files

File: checks/apps/cp2k/cp2k_check.py [hunk 6]
```diff
@@ -3,149 +3,208 @@
 #
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:3]: update the Copyright to 2016-2024

File: checks/apps/cp2k/cp2k_check.py [hunk 7]
```diff
@@ -3,149 +3,208 @@
...
+    """
+    Test CP2K build from source.
+    """
+
+    descr = "CP2K Build Test"
+    valid_systems = ["*"]
+    valid_prog_environs = ["+cp2k-dev"]
+    build_system = "CMake"
+    sourcesdir = None
+    maintainers = ["RMeli"]
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:60]:
```suggestion
    maintainers = ["SSA"]
```

File: checks/apps/cp2k/cp2k_check.py [hunk 8]
```diff
@@ -3,149 +3,208 @@
...
-    }
-
-    @run_after('init')
-    def set_prgenv(self):
-        if self.current_system.name in ['eiger', 'pilatus']:
-            self.valid_prog_environs = ['cpeGNU']
-        else:
-            self.valid_prog_environs = ['builtin']
+    executable = "./mps-wrapper.sh cp2k.psmp"
+    maintainers = ["RMeli"]
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:108]:
```suggestion
    maintainers = ["SSA"]
```

File: checks/apps/cp2k/cp2k_check.py [hunk 9]
```diff
@@ -3,149 +3,208 @@
...
+    sourcesdir = None
+    maintainers = ["RMeli"]
+    cp2k_sources = fixture(cp2k_download, scope="session")
+    build_locally = False
+
+    @run_before("compile")
+    def prepare_build(self):
+        self.uarch = uenv.uarch(self.current_partition)
+        self.build_system.builddir = os.path.join(self.stagedir, "build")
+        self.build_system.max_concurrency = 64
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:68]:
```suggestion
        self.skip_if_no_procinfo()
        cpu = self.current_partition.processor
        self.build_system.max_concurrency = cpu.info['num_cpus_per_socket']
```

**Comment by jgphpc** [comment]
CP2K is deployed on eiger:

```
uenv image find 
cp2k/2024.3:v1                          zen2  2024-09-18 7c7369b64b5fabe5 2.7GB     
cp2k/2024.1:v1                          zen2  2024-07-01 2a56f1df31a4c196 2.6GB  
```
hence we should either make the test work on eiger too or ignore eiger. What do you think ?

File: checks/apps/cp2k/cp2k_check.py [hunk 10]
```diff
@@ -3,149 +3,208 @@
...
+        # INFO: With -DCP2K_ENABLE_REGTEST=OFF, executables are in build/bin/ as expected
+        executable = os.path.join(self.stagedir, "exe", "local_cuda", "cp2k.psmp")
+        return os.path.isfile(executable)
 
 
 class Cp2kCheck(rfm.RunOnlyRegressionTest):
-    modules = ['CP2K']
-    executable = 'cp2k.psmp'
-    executable_opts = ['H2O-256.inp']
-    maintainers = ['LM']
```
[the remark this answers is not in the document]
Comment (jgphpc) [reply checks/apps/cp2k/cp2k_check.py:14]: see comment below

**Review by jgphpc — CHANGES_REQUESTED** [review]
see comments
(Tested with uenv/5.2.0-dev, UENV=cp2k/2024.3:v1 and 
[reframe.yaml](https://raw.githubusercontent.com/eth-cscs/alps-uenv/9a4367a350fb503b87c2afa7bdd7ec344f4a70cf/recipes/cp2k/2024.3/gh200/extra/reframe.yaml))

**Comment by RMeli** [comment]
Thanks @jgphpc for all the comments. I'll definitely add Eiger, and tackle the remaining TODO items.

**Comment by RMeli** [comment]
> we should either make the test work on eiger too or ignore eiger

It would be good to make the tests work on `eiger` too if the UENV is deployed there, but currently `eiger` has a different version of the SLURM plugin. Let's ignore it for now, and re-visit at a later stage.

**Review by teojgo — CHANGES_REQUESTED** [review]
My only comment would be to change to single quote strings when possible to be consistent with the other tests.

**Comment by RMeli** [comment]
Thanks, I'll do that. However, if we are interested in consistency down to quotes, I'd suggest to add a configuration file for some automated code formatter (eg `black`/`ruff`) so that everything is taken care of automagically.

File: checks/apps/cp2k/cp2k_check.py [hunk 11]
```diff
@@ -1,151 +1,230 @@
...
+    @run_before("run")
+    def prepare_run(self):
+        self.uarch = uenv.uarch(self.current_partition)
+
+        config = slurm_config[self.test_name][self.uarch]
+
+        # SBATCH options
+        self.job.options = [
+            f"--nodes={config['nodes']}",
+            "--ntasks-per-core=1",
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:137]:
```suggestion
    self.num_tasks_per_core = 1
```

File: checks/apps/cp2k/cp2k_check.py [hunk 12]
```diff
@@ -1,151 +1,230 @@
...
+            "--partition=debug",
```
Comment (jgphpc) [inline checks/apps/cp2k/cp2k_check.py:138]: This is not portable so I would remove it and call reframe with `-J partition=debug` (or add it in the config file)

Comment (RMeli) [reply checks/apps/cp2k/cp2k_check.py:138]: Yes, this is for me to test locally. I'll remove it before merging.

Comment (RMeli) [reply checks/apps/cp2k/cp2k_check.py]: Done, I added the debug partition to the configuration and removed any reference to the system from the tests. I then used `--system todi:debug` run the tests.

**Comment by RMeli** [comment]
I added tests for compilation with the `modules` view and updated the `reframe.yaml` in https://github.com/eth-cscs/alps-uenv/pull/156 accordingly: https://github.com/cp2k/cp2k/tree/support/v2024.3/benchmarks/QS_mp2_rpa/128-H2O

File: config/systems/todi.py [hunk 13]
```diff
@@ -76,6 +76,46 @@
                 ],
             'launcher': 'srun',
         },
+        {
+            'name': 'debug',
+            'descr': 'GH200',
+            'scheduler': 'squeue',
+            'time_limit': '10m',
+            'container_platforms': [],
+            'access': ['--partition=debug'],
```
Comment (RMeli) [inline config/systems/todi.py:85]: Needs to be removed?

File: config/systems/todi.py [hunk 14]
```diff
@@ -76,6 +76,46 @@
                 ],
             'launcher': 'srun',
         },
+        {
+            'name': 'debug',
+            'descr': 'GH200',
+            'scheduler': 'squeue',
```
Comment (RMeli) [inline config/systems/todi.py:82]:
```suggestion
            'scheduler': 'slurm',
```
Using `squeue` circumvents `sacct` issues on `todi-ln001`.

File: config/systems/todi.py (hunk 13 above)
Comment (teojgo) [reply config/systems/todi.py:85]: Yes, I would say remove it for the moment and we can add it later in case we need.

**Review by teojgo — CHANGES_REQUESTED** [review]

**Comment by RMeli** [comment]
I added an additional short MD test, and tested the whole pipeline with `'access': ['--reservation=dain']`. Everything completed successfully.

**Review by teojgo — APPROVED** [review]

**Review by ekouts — APPROVED** [review]

**Review by jgphpc — APPROVED** [review]
lgetm

**Pull request merged**